app/api/services/EnrolmentsAuthService.scala

```diff
@@ -39,55 +38,51 @@ class EnrolmentsAuthService @Inject() (val connector: AuthConnector, val appConf
     } else {
       predicate
     }
 
   def authorised(predicate: Predicate)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[AuthOutcome] = {
-    authFunction.authorised(buildPredicate(predicate)).retrieve(affinityGroup and authorisedEnrolments) {
-      case Some(Individual) ~ _ =>
-        val user = UserDetails("", "Individual", None)
-        Future.successful(Right(user))
-      case Some(Organisation) ~ _ =>
-        val user = UserDetails("", "Organisation", None)
-        Future.successful(Right(user))
-      case Some(Agent) ~ _ =>
-        retrieveAgentDetails() map {
-          case arn @ Some(_) =>
-            val user: AuthOutcome = Right(UserDetails("", "Agent", arn))
-            user
-          case None =>
-            logger.warn(s"[EnrolmentsAuthService][authorised] No AgentReferenceNumber defined on agent enrolment.")
-            Left(InternalError)
-        }
-      case _ ~ _ =>
-        logger.warn(s"[EnrolmentsAuthService][authorised] Invalid AffinityGroup.")
-        Future.successful(Left(ClientNotAuthorisedError))
-    } recoverWith {
-      case _: MissingBearerToken     => Future.successful(Left(ClientNotAuthorisedError))
-      case _: AuthorisationException => Future.successful(Left(ClientNotAuthorisedError))
-      case error =>
-        logger.warn(s"[EnrolmentsAuthService][authorised] An unexpected error occurred: $error")
-        Future.successful(Left(InternalError))
-    }
+    authFunction
+      .authorised(buildPredicate(predicate))
+      .retrieve(affinityGroup) {
+        case Some(Individual)   => Future.successful(Right(UserDetails("", "Individual", None)))
+        case Some(Organisation) => Future.successful(Right(UserDetails("", "Organisation", None)))
+        case Some(Agent) =>
+          retrieveAgentDetails().map {
+            case arn @ Some(_) => Right(UserDetails("", "Agent", arn))
+            case None =>
+              logger.warn(s"[EnrolmentsAuthService][authorised] No AgentReferenceNumber defined on agent enrolment.")
+              Left(InternalError)
+          }
+        case _ =>
+          logger.warn(s"[EnrolmentsAuthService][authorised] Invalid AffinityGroup.")
+          Future.successful(Left(ClientOrAgentNotAuthorisedError))
+      }
+      .recoverWith {
+        case _: MissingBearerToken     => Future.successful(Left(ClientOrAgentNotAuthorisedError))
+        case _: AuthorisationException => Future.successful(Left(ClientOrAgentNotAuthorisedError))
+        case error =>
+          logger.warn(s"[EnrolmentsAuthService][authorised] An unexpected error occurred: $error")
+          Future.successful(Left(InternalError))
+      }
   }
 
-  private def retrieveAgentDetails()(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[String]] =
+  private def retrieveAgentDetails()(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[String]] = {
+    def getAgentReferenceFrom(enrolments: Enrolments) =
```

Comment:
I think we should add the return type for the method as `Option[String]` to make it clearer?

Reply:
Hmm. Maybe. But it's a local method inside a private method and this is shared code so should be changed in BSAS if you feel strongly about it (otherwise it will get overwritten).